File: old/actions/configure_postgres_files.py
"""
Configure PostgreSQL files action
"""
import logging
from cli import FileOps, Sed
from .base import Action
logger = logging.getLogger(__name__)

class ConfigurePostgresFilesAction(Action):
    """Action to configure PostgreSQL configuration files"""
    description = "postgresql files configuration"

    def execute(self) -> bool:
        """Configure PostgreSQL configuration files"""
        if not self.ssh_service:
            logger.error("SSH service not initialized")
            return False
        # Get params
        version = "17"
        port = 5432
        # Use network from config (environment-specific) if available, otherwise fallback to container params or default
        allow_cidr = "10.11.3.0/24"
        if self.cfg and hasattr(self.cfg, "network") and self.cfg.network:
            allow_cidr = self.cfg.network
        if self.container_cfg:
            params = self.container_cfg.params or {}
            version = str(params.get("version", "17"))
            port = params.get("port", 5432)
            # Only use container param cidr if config network is not available
            if not (self.cfg and hasattr(self.cfg, "network") and self.cfg.network):
                allow_cidr = params.get("cidr", allow_cidr)
        config_path = f"/etc/postgresql/{version}/main/postgresql.conf"
        # Remove all existing listen_addresses lines
        remove_cmd = f"sed -i '/^#*listen_addresses.*/d' {config_path}"
        output, exit_code = self.ssh_service.execute(remove_cmd, sudo=True)
        if exit_code is not None and exit_code != 0:
            logger.error("remove listen_addresses failed with exit code %s", exit_code)
            return False
        # Check if listen_addresses already exists (shouldn't after removal, but check anyway)
        check_cmd = f"grep -q '^listen_addresses =' {config_path} && echo exists || echo not_exists"
        check_output, _ = self.ssh_service.execute(check_cmd, sudo=True)
        if "not_exists" in check_output:
            # Find the line number of CONNECTIONS section
            line_num_cmd = f"grep -n '^# CONNECTIONS AND AUTHENTICATION' {config_path} | cut -d: -f1"
            line_num_output, _ = self.ssh_service.execute(line_num_cmd, sudo=True)
            if line_num_output and line_num_output.strip().isdigit():
                line_num = int(line_num_output.strip())
                # Use awk to insert the line after the specified line number
                insert_script = f"""awk -v n={line_num} -v s="listen_addresses = '*'" 'NR==n {{print; print s; next}}1' {config_path} > {config_path}.tmp && mv {config_path}.tmp {config_path}
"""
            else:
                # Fallback: append at end of file
                insert_script = f"""echo "listen_addresses = '*'" >> {config_path}
"""
            output, exit_code = self.ssh_service.execute(insert_script, sudo=True)
            if exit_code is not None and exit_code != 0:
                logger.error("insert listen_addresses failed with exit code %s", exit_code)
                if output:
                    logger.error("insert listen_addresses output: %s", output)
                return False
        port_cmd = Sed().flags("").replace(config_path, "^#?port.*", f"port = {port}")
        # Add both local and network entries to pg_hba.conf
        pg_hba_content = f"local all all peer\nhost all all {allow_cidr} md5\n"
        pg_hba_cmd = FileOps().write(f"/etc/postgresql/{version}/main/pg_hba.conf", pg_hba_content)
        results = []
        for cmd, desc in [
            (port_cmd, "configure postgres port"),
            (pg_hba_cmd, "write pg_hba rule"),
        ]:
            output, exit_code = self.ssh_service.execute(cmd, sudo=True)
            if exit_code is not None and exit_code != 0:
                logger.error("%s failed with exit code %s", desc, exit_code)
                if output:
                    logger.error("%s output: %s", desc, output.splitlines()[-1])
            results.append(exit_code is None or exit_code == 0)
        # Restart PostgreSQL to apply listen_addresses change
        if all(results):
            logger.info("Restarting PostgreSQL to apply configuration changes...")
            from cli import SystemCtl
            import time
            cluster_service = f"postgresql@{version}-main"
            restart_cmd = SystemCtl().service(cluster_service).restart()
            output, exit_code = self.ssh_service.execute(restart_cmd, sudo=True)
            if exit_code is not None and exit_code != 0:
                logger.error("restart postgresql failed with exit code %s", exit_code)
                return False
            time.sleep(3)
            # Verify it's listening on all interfaces
            port_check_cmd = "ss -tlnp | grep :5432 | grep -v '127.0.0.1' || echo 'not_listening'"
            port_output, _ = self.ssh_service.execute(port_check_cmd, sudo=True)
            if "not_listening" not in port_output and ":5432" in port_output:
                logger.info("PostgreSQL is listening on all interfaces after restart")
            else:
                logger.warning("PostgreSQL may not be listening on external interfaces after restart")
        return all(results)

File: src/actions/configure_postgres_files.go
package actions

import (
	"fmt"
	"strings"
	"time"
	"enva/cli"
	"enva/libs"
	"enva/services"
)

// ConfigurePostgresFilesAction configures PostgreSQL configuration files
type ConfigurePostgresFilesAction struct {
	*BaseAction
}

func NewConfigurePostgresFilesAction(sshService *services.SSHService, aptService *services.APTService, pctService *services.PCTService, containerID *string, cfg *libs.LabConfig, containerCfg *libs.ContainerConfig) Action {
	return &ConfigurePostgresFilesAction{
		BaseAction: &BaseAction{
			SSHService:   sshService,
			APTService:   aptService,
			PCTService:   pctService,
			ContainerID: containerID,
			Cfg:          cfg,
			ContainerCfg: containerCfg,
		},
	}
}

func (a *ConfigurePostgresFilesAction) Description() string {
	return "postgresql files configuration"
}

func (a *ConfigurePostgresFilesAction) Execute() bool {
	if a.SSHService == nil {
		libs.GetLogger("configure_postgres_files").Error("SSH service not initialized")
		return false
	}
	version := "17"
	port := 5432
	allowCIDR := "10.11.3.0/24"
	if a.Cfg != nil && a.Cfg.Network != "" {
		allowCIDR = a.Cfg.Network
	}
	if a.ContainerCfg != nil && a.ContainerCfg.Params != nil {
		if v, ok := a.ContainerCfg.Params["version"].(string); ok {
			version = v
		}
		if p, ok := a.ContainerCfg.Params["port"].(int); ok {
			port = p
		}
		if a.Cfg == nil || a.Cfg.Network == "" {
			if cidr, ok := a.ContainerCfg.Params["cidr"].(string); ok {
				allowCIDR = cidr
			}
		}
	}
	configPath := fmt.Sprintf("/etc/postgresql/%s/main/postgresql.conf", version)
	// Remove all existing listen_addresses lines (matching Python: sudo=True)
	removeCmd := fmt.Sprintf("sed -i '/^#*listen_addresses.*/d' %s", configPath)
	output, exitCode := a.SSHService.Execute(removeCmd, nil, true) // sudo=True
	if exitCode != nil && *exitCode != 0 {
		libs.GetLogger("configure_postgres_files").Error("remove listen_addresses failed with exit code %d", *exitCode)
		return false
	}
	// Check if listen_addresses already exists (matching Python: sudo=True)
	checkCmd := fmt.Sprintf("grep -q '^listen_addresses =' %s && echo exists || echo not_exists", configPath)
	checkOutput, _ := a.SSHService.Execute(checkCmd, nil, true) // sudo=True
	if strings.Contains(checkOutput, "not_exists") {
		// Find the line number of CONNECTIONS section (matching Python: sudo=True)
		lineNumCmd := fmt.Sprintf("grep -n '^# CONNECTIONS AND AUTHENTICATION' %s | cut -d: -f1", configPath)
		lineNumOutput, _ := a.SSHService.Execute(lineNumCmd, nil, true) // sudo=True
		if lineNumOutput != "" && strings.TrimSpace(lineNumOutput) != "" {
			insertScript := fmt.Sprintf("awk -v n=%s -v s=\"listen_addresses = '*'\" 'NR==n {print; print s; next}1' %s > %s.tmp && mv %s.tmp %s", strings.TrimSpace(lineNumOutput), configPath, configPath, configPath, configPath)
			output, exitCode = a.SSHService.Execute(insertScript, nil, true) // sudo=True
		} else {
			// Fallback: append at end of file (matching Python: sudo=True)
			insertScript := fmt.Sprintf("echo \"listen_addresses = '*'\" >> %s", configPath)
			output, exitCode = a.SSHService.Execute(insertScript, nil, true) // sudo=True
		}
		if exitCode != nil && *exitCode != 0 {
			libs.GetLogger("configure_postgres_files").Error("insert listen_addresses failed with exit code %d", *exitCode)
			if output != "" {
				libs.GetLogger("configure_postgres_files").Error("insert listen_addresses output: %s", output)
			}
			return false
		}
	}
	portCmd := cli.NewSed().Flags("").Replace(configPath, "^#?port.*", fmt.Sprintf("port = %d", port))
	pgHbaContent := fmt.Sprintf("local all all peer\nhost all all %s md5\n", allowCIDR)
	pgHbaCmd := cli.NewFileOps().Write(fmt.Sprintf("/etc/postgresql/%s/main/pg_hba.conf", version), pgHbaContent)
	results := []bool{}
	for _, item := range []struct {
		cmd string
		desc string
	}{
		{portCmd, "configure postgres port"},
		{pgHbaCmd, "write pg_hba rule"},
	} {
		// Matching Python: sudo=True for all commands
		output, exitCode := a.SSHService.Execute(item.cmd, nil, true) // sudo=True
		if exitCode != nil && *exitCode != 0 {
			libs.GetLogger("configure_postgres_files").Error("%s failed with exit code %d", item.desc, *exitCode)
			if output != "" {
				lines := strings.Split(output, "\n")
				if len(lines) > 0 {
					libs.GetLogger("configure_postgres_files").Error("%s output: %s", item.desc, lines[len(lines)-1])
				}
			}
		}
		results = append(results, exitCode == nil || *exitCode == 0)
	}
	if results[0] && results[1] {
		libs.GetLogger("configure_postgres_files").Info("Restarting PostgreSQL to apply configuration changes...")
		clusterService := fmt.Sprintf("postgresql@%s-main", version)
		restartCmd := cli.NewSystemCtl().Service(clusterService).Restart()
		output, exitCode = a.SSHService.Execute(restartCmd, nil, true) // sudo=True
		if exitCode != nil && *exitCode != 0 {
			libs.GetLogger("configure_postgres_files").Error("restart postgresql failed with exit code %d", *exitCode)
			return false
		}
		time.Sleep(3 * time.Second)
		// Verify it's listening on all interfaces (matching Python: sudo=True)
		portCheckCmd := "ss -tlnp | grep :5432 | grep -v '127.0.0.1' || echo 'not_listening'"
		portOutput, _ := a.SSHService.Execute(portCheckCmd, nil, true) // sudo=True
		if !strings.Contains(portOutput, "not_listening") && strings.Contains(portOutput, ":5432") {
			libs.GetLogger("configure_postgres_files").Info("PostgreSQL is listening on all interfaces after restart")
		} else {
			libs.GetLogger("configure_postgres_files").Warning("PostgreSQL may not be listening on external interfaces after restart")
		}
	}
	return results[0] && results[1]
}

